### PR TITLE
JDK-8291974 PrivateCredentialPermission should not use local variable to enable debugging

### PR DESCRIPTION
**Pull request by mcpowers**
https://bugs.openjdk.org/browse/JDK-8291974

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8291974](https://bugs.openjdk.org/browse/JDK-8291974): PrivateCredentialPermission should not use local variable to enable debugging
 * [JDK-8293653](https://bugs.openjdk.org/browse/JDK-8293653): PrivateCredentialPermission should not use local variable to enable debugging (**CSR**)

### Reviewers
 * [Sean Mullan](https://openjdk.org/census#mullan) (@seanjmullan - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10206/head:pull/10206` \
`$ git checkout pull/10206`

Update a local copy of the PR: \
`$ git checkout pull/10206` \
`$ git pull https://git.openjdk.org/jdk pull/10206/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10206`

View PR using the GUI difftool: \
`$ git pr show -t 10206`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10206.diff">https://git.openjdk.org/jdk/pull/10206.diff</a>

</details>

### CODE REVIEW

**Comment by mcpowers** [comment]
What happens when deserialization encounters a missing field like `testing`? Does it ignore it?

**Comment by seanjmullan** [comment]
I would write a test which serializes the data (before your change) and deserializes it after. There should be some regression tests that already do that.

**Comment by mcpowers** [comment]
@seanjmullan There is an existing test which serializes and deserializes a `PrivateCredentialPermission` object. It only checks that the deserialized object is same as the original. There is no testing for new or deleted fields. The Java Serialization Spec indicates that the `testing` variable can be safely removed. This is only possible because the variable is `boolean` and is initialized to `false`. The CSR has more details. I have verified this behavior with a manual test of my own.

Are you suggesting that I write and integrate a new test that basically verifies that the Java Serialization Spec works as documented?

**Comment by valeriepeng** [comment]
> What happens when deserialization encounters a missing field like `testing`? Does it ignore it?

Have we tried this to know for sure that it is ignored? I think you should verify this by serializing w/o this "testing" field and de-serialize it with older releases.

**Comment by mcpowers** [comment]
I wrote a test to deserialize an object created and serialized before the fix. I compare the old object with the current object after the fix. What is important is not the comparison, but that deserialization doesn't throw an exception.

**Comment by valeriepeng** [comment]
> I wrote a test to deserialize an object created and serialized before the fix. I compare the old object with the current object after the fix. What is important is not the comparison, but that deserialization doesn't throw an exception.

Does both de-serialization pass, i.e. deserialize the new bytes using old release and deserializing the old bytes using new release?

**Comment by mcpowers** [comment]
I tested old bytes on new release. It passes. I didn't test new bytes on an old release because of Max's "from the future" comment. Honestly, I'm not sure what any of this testing proves. Maybe that the Java Object Serialization Specification isn't broken?

**Comment by mcpowers** [comment]
Fair enough. I'll run a test in the opposite direction.

**Comment by valeriepeng** [comment]
> I tested old bytes on new release. It passes. I didn't test new bytes on an old release because of Max's "from the future" comment. Honestly, I'm not sure what any of this testing proves. Maybe that the Java Object Serialization Specification isn't broken?

Supposedly if the newer bytes doesn't work with the older releases, it is a compatibility format change and you should change the serialVersionID value to indicate it. In this case, some may wonder if it's worthwhile to remove this field.

**Comment by RogerRiggs** [comment]
As mentioned earlier in the comments...
This is a completely compatible change.  Removing a field is well specified as a compatible change.
From old version to new version the value in the stream is discarded.
From a new version to the old version, the value in the stream is missing so the field is initialized to its default value (false for boolean).
There's no reason to change the serialVersionUID. Changing it would be an incompatible change and cause all kinds of other problems. 

In the current case, since the value of the field is false, it will be hard to observe any change in behavior.

**Comment by mcpowers** [comment]
I ran the test in the reverse direction. No difference in behavior was observed.

**Comment by valeriepeng** [comment]
> I ran the test in the reverse direction. No difference in behavior was observed.

Great, thanks for confirming.

**Comment by valeriepeng** [comment]
> As mentioned earlier in the comments... This is a completely compatible change. Removing a field is well specified as a compatible change. From old version to new version the value in the stream is discarded. From a new version to the old version, the value in the stream is missing so the field is initialized to its default value (false for boolean). There's no reason to change the serialVersionUID. Changing it would be an incompatible change and cause all kinds of other problems.
> 
> In the current case, since the value of the field is false, it will be hard to observe any change in behavior.

Thanks for the comments, just want to be sure the old version handles the missing field, that's all.

File: test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import java.io.*;
+import java.util.*;
+
+public class Serial2 implements java.io.Serializable {
+
+    /*
+     * Base64 encoding of Serialized PrivateCredentialPermission object
+     * before bug fix for JDK-8291974.
+     */
+    static String before = "rO0ABXNyAC9qYXZheC5zZWN1cml0eS5hdXRoLlByaXZhdGVDcmVkZW50aWFsUGVybWlzc2lvbklV3Hd7UH9MAgADWgAHdGVzdGluZ0wAD2NyZWRlbnRpYWxDbGFzc3QAEkxqYXZhL2xhbmcvU3RyaW5nO0wACnByaW5jaXBhbHN0AA9MamF2YS91dGlsL1NldDt4cgAYamF2YS5zZWN1cml0eS5QZXJtaXNzaW9uscbhPyhXUX4CAAFMAARuYW1lcQB+AAF4cHQAGWNyZWQxIHBjMSAicG4xIiBwYzIgInBuMiIAdAAFY3JlZDFw";
```
[the remark this answers is not in the document]
Comment (mcpowers) [reply test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java:42]: Or use string concatenation, e.g. "xxx" + "yyy".

[the remark this answers is not in the document]
Comment (seanjmullan) [reply test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java:42]: This looks like a good candidate for Text Blocks: https://openjdk.org/jeps/378

[the remark this answers is not in the document]
Comment (mcpowers) [reply test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java:42]:
```
        String a = """
            xxx
            yyy
        """;
```

is not the same as

```
        String a =
            "xxx" +
            "yyy"
        ;
```
`a` contains an embedded newline; `b` does not.

[the remark this answers is not in the document]
Comment (efge) [reply test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java:42]: @mcpowers the idea would be to use

    String a = """
            xxx\
            yyy\
        """;

[the remark this answers is not in the document]
Comment (mcpowers) [reply test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java:42]: I'll go with Text Blocks and Florent's suggestion, but thanks for mentioning `getMimeDecoder`. I'm sure I'll find a use for it in the future.

File: test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        try {
+            // Decode Base64 string and turn it into an input stream.
+            byte[] decoded = Base64.getDecoder().decode(before);
+            InputStream is = new ByteArrayInputStream(decoded);
+
+            // Deserialize input stream and create a new object.
+            ObjectInputStream ois = new ObjectInputStream(is);
+            PrivateCredentialPermission pcp2 =
+                    (PrivateCredentialPermission)ois.readObject();
+            is.close();
```
Comment (seanjmullan) [inline test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java:61]: Not necessary as BAIS.close() is a no-op.

File: test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    public static void main(String[] args) {
+
+        try {
+            // Decode Base64 string and turn it into an input stream.
+            byte[] decoded = Base64.getDecoder().decode(before);
+            InputStream is = new ByteArrayInputStream(decoded);
+
+            // Deserialize input stream and create a new object.
+            ObjectInputStream ois = new ObjectInputStream(is);
```
Comment (seanjmullan) [inline test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java:58]: Use try-with resources here so the input stream is automatically closed even if there are exceptions.

File: test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+             */
+            if (!pcp.equals(pcp2) || !pcp2.equals(pcp)) {
+                throw new SecurityException("Serial2 test failed: " +
+                                        "EQUALS TEST FAILED");
+            }
+
+            System.out.println("Serial2 test succeeded");
+        } catch (Exception e) {
+            e.printStackTrace();
+            throw new SecurityException("Serial test failed");
```
Comment (seanjmullan) [inline test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java:79]: If you include e as the cause (2nd argument) of `SecurityException` then you don't need to print the stack trace on line 78.

File: test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8291974
+ * @summary     PrivateCredentialPermission should not use local variable to enable debugging
+ *              implementation-dependent class
+ * @run main/othervm/policy=Serial.policy Serial2
```
Comment (seanjmullan) [inline test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java:29]: I think you can remove `main/othervm/policy=Serial.policy` (actually you can remove the whole `@run` line then).
The policy argument causes the test to run with a SecurityManager enabled, and there isn't any reason that this test needs to do that AFAICT. Also that policy file is for other tests in this directory for accessing the file system or JAAS credentials, which you are not accessing in this test.

File: test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+ * @bug 8291974
+ * @summary     PrivateCredentialPermission should not use local variable to enable debugging
+ *              implementation-dependent class
+ */
+
+import javax.security.auth.*;
+import java.io.*;
+import java.util.*;
+
+public class Serial2 implements java.io.Serializable {
```
Comment (seanjmullan) [inline test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java:35]: I don't see why this needs to implement `Serializable` (I see you probably used the other `Serial.java` test in this directory as a template. I don't think that needs it either, but you can leave that as is for now).

File: test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java (hunk 2 above)
Comment (mcpowers) [reply test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java]: removed

File: test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java (hunk 3 above)
Comment (mcpowers) [reply test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java]: using try-with resources

File: test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java (hunk 4 above)
Comment (mcpowers) [reply test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java]: fixed

File: test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java (hunk 5 above)
Comment (mcpowers) [reply test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java]: fixed

File: test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java (hunk 6 above)
Comment (mcpowers) [reply test/jdk/javax/security/auth/PrivateCredentialPermission/Serial2.java:35]: I read your sentence as "leave both for now". Sorry. You mean fix in this PR and leave the other alone.

**Comment by seanjmullan** [comment]
Looks good. Go ahead an integrate and I will sponsor.

**Review by seanjmullan — APPROVED** [review]

**Pull request closed without merging**